### PR TITLE
added a trace to the serverside room

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
When enabled, a trace file will be created containing every websocket broadcast the room sends and a few other things.

The trace can be enabled for a room by pressing `F9`.

This is a very short and simple trace file I recorded: [trace.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/6542338/trace.zip) (extract the file)

You can press `CTRL+F9` to load the file and you should get a slider on top to display the room state at different times. Use a browser window in portrait orientation so it's not obscured.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3095/pr-test (or any other room on that server)

File: server/room.mjs [hunk 1]
```diff
@@ -279,12 +281,20 @@ export default class Room {
...
     const meta = this.state._meta;
     this.state = state;
     this.state._meta = meta;
     this.broadcast('state', state);
   }
 
+  trace(type, payload) {
+    if(this.traceActivated) {
+      const line = `${new Date().toISOString()} - ${type} - ${JSON.stringify(payload)}\n`;
+      fs.appendFileSync(path.resolve() + '/save/' + this.id + '.trace', line);
```
Comment (UltimateGeek) [inline server/room.mjs:294]: Consider sanitizing `this.id` using something like this: 

https://www.npmjs.com/package/string-sanitizer

**Comment by 96LawDawg** [comment]
I tried to test this.  I thought I figured out how to activate the log, but have no idea how to review it.

**Comment by ArnoldSmith86** [comment]
Yeah, there's no way for you to review the log. It is only on the server.

**Comment by RaphaelAlvez** [comment]
@ArnoldSmith86 how should we review this PR. I think @UltimateGeek is not helping at the moment.

**Comment by rogerl50** [comment]
A couple of questions: 
1. It looks like this replaces the logging function. That function was used only in the `window.onerror` handler in `main.js`. Should the call to log there be replaced by an explicit call to `console.log`? As the PR stands now, that error (which I presume is a serious error) will be lost if tracing is not enabled (I think).
2. Do people think that this set of trace points is likely to catch whatever the timing issue is that seems to cause delays in moving cards during busy times in the room?

File: server/room.mjs (hunk 1 above)
Comment (ArnoldSmith86) [reply server/room.mjs:339]: In the meantime, server.mjs takes care of only accepting `/^[A-Za-z0-9_-]+$/` as room IDs.

**Comment by ArnoldSmith86** [comment]
Here is me playing a game of FreeCell: [98r2-1622061576178.trace.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/6549596/98r2-1622061576178.trace.zip)

**Comment by 96LawDawg** [comment]
Looking at the code is as much fun as watching paint dry.  You have a commit that says there is a replay viewer.  How does that work?

**Comment by ArnoldSmith86** [comment]
You currently can't record a trace and then play it back because you need access to the created file and that's only on the server. I'll probably add a way to do it though.

You can download one of the two zip files in my comments, extract them, press `CTRL-F9` and select the trace file.

**Comment by ArnoldSmith86** [comment]
>     1. It looks like this replaces the logging function. That function was used only in the `window.onerror` handler in `main.js`. Should the call to log there be replaced by an explicit call to `console.log`? As the PR stands now, that error (which I presume is a serious error) will be lost if tracing is not enabled (I think).

Yes. I imagine it will be 1000x more useful to have the complete context when an error happens. I don't really want to go through error messages in the general log.

>     2. Do people think that this set of trace points is likely to catch whatever the timing issue is that seems to cause delays in moving cards during busy times in the room?

Unlikely I guess. This is more for finding unnecessary conflict resolutions or bugs like one card in the hand being flipped etc.

**Comment by ArnoldSmith86** [comment]
Yay, the automated check `Production Environment` saved the day on this one! :smile:

So, this isn't exactly "finished" but it is already very useful. It was in 4 staging PRs already and didn't cause problems.

I would like to get this merged soonish (no hurry) because it's quite a hassle to have actual game nights on a PR test server. Typing IP addresses isn't fun and some people straight up can't join if the port isn't 443. But actual game nights are what this PR is made for. So you can enable tracing at the start and just press `F9` whenever a bug occurs without having to care about it while actually playing.

**Review by 96LawDawg — APPROVED** [review]
I've seen the results and it doesn't seem to cause any problems.  Approved through test.

**Review by RaphaelAlvez — APPROVED** [review]
Tested and approved. There is still room for improvement in the output but I think this should be merged as someone as possible so we can get better bug report

**Pull request merged**